### DISCUSSION TITLE
How to deal with a host that has z-levels defined continuously from ocean to atmosphere

### DISCUSSION THREAD

**Discussion by platipodium** [fabm-model/fabm#50] [Q&A]
Dear community,

I have a host model that has z-levels defined at interfaces, e.g., as `(/10,9,8,5,2,1,0,-10,-10000/)'  i.e. in the ocean 10m, 9 m and so forth and -10 for 10 m above surface and -10000 for 10 km above surface.   The water surface elevation is then calculated in addition either as part of the layer from depth 1 m to 0 m or as part of the layer from depth 0 to -10.  
 * I added to the host the surface_index()  analogously to bottom_index() defined in other hosts to point to the maximum index in the z-levels that is still (partly) water.
 * for all wet vertical levels, I update cell_thickness each timestep, I set cell_thickness to 0 for atmosphere
 * I tried running gotm/light original, but (... see error below)
 * I tampered with gotm/light to calculate the light profile starting not from z=nk to bottom_index(), but from surface_index() to bottom_index().  Still I get the following error:
```
Fatal ERROR at check_vertical_call_output:
 /light:do_column wrote non-finite data for light/swr
```

Any ideas on fixing this, or a better way to deal with variable surface_index?

**Comment by platipodium** [comment]
This vertical loop is working, but I still wonder whether there's a better (more fabm-generic) way?

```
  _DOWNWARD_LOOP_BEGIN_
         _GET_(self%id_dz,dz)         ! Layer height (m)
         _GET_(self%id_ext,ext)       ! PAR attenuation (m-1)
         _GET_(self%id_depth,depth)   ! Layer center depth

         ! Calculate photosynthetically active radiation (PAR), shortwave radiation, and PAR attenuation.
         ! only do this if in water (depth>0)
         if (depth > 0) then 
           z = z + depth * 0.5_rk
           bioext = bioext + ext * depth * 0.5_rk
           par = swr0 * (1.0_rk - self%a) * exp(-z / self%g2 - bioext)
           swr = par + swr0 * self%a * exp(-z / self%g1)
           ! Move to bottom of layer
           z = z + depth * 0.5_rk
           bioext = bioext + ext * depth * 0.5_rk
           !write(*,*) __LINE__,swr,par,z,bioext
         else ! we're in the atmosphere, so keep light constant
           par = swr0
           swr = par
         endif

         _SET_DIAGNOSTIC_(self%id_swr,swr) ! Shortwave radiation at layer centre
         _SET_DIAGNOSTIC_(self%id_par,par) ! Photosynthetically active radiation at layer centre

      _DOWNWARD_LOOP_END_
   ```

**Comment by platipodium** [comment]
But even if the do_column is running fine, I prepare_inputs() exits with a floating overflow...

**Comment by jornbr** [comment]
Hi Carsten,

I assume this model has more dimensions than depth alone? If so, can the surface index can vary in the horizontal (as well as in time)? I'd aim for a solution where non-water points are filtered out by FABM before they reach your `do`/`do_surface`/`do_bottom`/`do_column` routines. Since there is currently no surface equivalent of `set_bottom_index`, I'd look instead towards using `set_domain_start` to specify the first index for each dimension, e.g., `call model%set_domain_start(1,1,ksurf)` for a 3D model. Here, `ksurf` must be a scalar, and thus the same in every horizontal point. If that is not appropriate, I would additionally provide FABM with a mask (perhaps you are doing so already to indicate land points?). In that mask, near-surface cells can then be masked out whenever they become dry. The end result will be that any internal routine (e.g., `do_column` in the light module) will only see the water points, so gotm/light and the like should work out of the box.
